source/docs/software/advanced-gradlerio/profiling-with-visualvm.rst

```diff
@@ -0,0 +1,176 @@
+
+This should launch VisualVM. Once launched, open the :guilabel:`Add JMX Connection` dialog.
+
+.. image:: images/visualvm/visualvm-addconn.png
+   :alt: Add visualvm connection menu option
+   :width: 700
+
+Once opened, configure the connection details and ensure that :guilabel:`Do not require SSL connection` is ticked.
+
+.. image:: images/visualvm/visualvm-dialog.png
+   :alt: VisualVM connection dialog is ticked
+   :width: 700
+
+If correctly done, a new menu option in the left-hand sidebar will appear. Clicking on it will show you a detailed dashboard of the running JVM application.
+
+.. image:: images/visualvm/visualvm-dash.png
+   :alt: VisualVM diagnostics dashboard
+   :width: 700
+
+Analyzing Function Timings
+--------------------------
+
+An important feature of VisualVM is the ability to view how much time a specific function is taking up. This is *without* having a code debugger attached. To begin, click on the :guilabel:`Sampler` tab and then click on :guilabel:`CPU`. This will immediately give a breakdown of what functions are taking CPU time.
+
+.. image:: images/visualvm/visualvm-function-profiling.png
+   :alt: Analyzing the VisualVM function timing tree
+   :width: 700
+
+The above screenshot shows a breakdown of the total time a specific function takes. You can see that ``totallyNotSlowFunction()`` accounts for ``61.9%`` of the robot program CPU time. We can then correlate this to our robot program. In ``totallyNotSlowFunction()``, we have the following code.
+
+.. code-block:: Java
+
+   public static void totallyNotSlowFunction() {
+      for (int i = 0; i < 2000; i++) {
+         System.out.println("HAHAHAHA");
+      }
+   }
+
+In this code snippet, we can identify 2 major causes of concern. A long running ``for`` loop blocks the rest of the robot program from running. Additionally, ``System.out.println()`` calls on the roboRIO are typically quite expensive. We found this information by profiling the Java application on the roboRIO!
+
+Creating a Heap Dump
+--------------------
+
+Besides viewing the remote systems CPU and memory usage, VisualVM is most useful by creating a **Heap Dump**. When a Java object is created, it resides in an area of memory called the heap. When the heap is full, a process called `garbage collection <https://www.geeksforgeeks.org/garbage-collection-java/>`__ begins. Garbage collection can be a common cause of loop overruns in a traditional Java robot program.
+
+To begin, ensure you are on the :guilabel:`Monitor` tab and click :guilabel:`Heap Dump`.
+
+.. image:: images/visualvm/visualvm-perform-heapdump.png
+   :alt: Location of heap dump button in VisualVM
+   :width: 700
+
+This heap dump will be stored on the target system (roboRIO) and must be retrieved using SFTP. For this tutorial, `FileZilla <https://filezilla-project.org/>`__ will be the tool of choice. Open FileZilla and connect to the roboRIO with the following parameters
```

Comment:
This could link to the existing doc for getting files off. https://frc-docs--2251.org.readthedocs.build/en/2251/docs/software/roborio-info/roborio-ftp.html